### DISCUSSION TITLE
Type problem with persist

### DISCUSSION THREAD

**Discussion by nerkn** [pmndrs/zustand#2176] [Q&A] [answered]
## Summary
Code I've written for 4.4.4  which is working very fine,  gave typescript error on 4.4.5

argument of type 'StateCreator<useCartType, [], [["zustand/persist", useCartType]]>' is not assignable to parameter of type 'StateCreator<useCartType, [], [never, unknown][]>'. Type 'StateCreator<useCartType, [], [["zustand/persist", useCartType]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'. Types of property '$$storeMutators' are incompatible. Type '[["zustand/persist", useCartType]] | undefined' is not assignable to type '[never, unknown][] | undefined'. Type '[["zustand/persist", useCartType]]' is not assignable to type '[never, unknown][]'. Type '["zustand/persist", useCartType]' is not assignable to type '[never, unknown]'. Type at position 0 in source is not compatible with type at position 0 in target. Type 'string' is not assignable to type 'never'.

## Link to reproduction

[https://github.com/nerkn/ecommerce-Frontend/blob/main/src/hooks/useCart.ts
](https://github.com/nerkn/ecommerce-Frontend/blob/main/src/hooks/useCart.ts)

## Check List
 
- [x] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.
comment of  muammed

[https://stackoverflow.com/questions/77402109/zustand-and-typescript-persist-function-cannot-be-implemented](https://stackoverflow.com/questions/77402109/zustand-and-typescript-persist-function-cannot-be-implemented)

Please include a minimal reproduction.

- [ ] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [x] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by jwindhager** [comment]
Can concur, experiencing the same issue with using the `immer` middleware (immer@10.0.3) in v4.4.5. No problem with v4.4.4.

**Comment by alexanderhorner** [comment]
I was about to submit this on Tuesday but couldn't reproduce it. Since wednesday was holiday, I just got around to it. What a coincidence that you beat me by 39 minutes.

Here is my reproduction:
https://codesandbox.io/p/sandbox/vibrant-bash-prcf7q?file=%2Fsrc%2Fstore.ts%3A21%2C6

It only happens with `"moduleResolution": "bundler",` in tsconfig.

**Comment by eisa-rezaei** [comment]
I have the same issue

**Comment by dbritto-dev** [comment, marked as answer]
@nerkn @jwindhager  @alexanderhorner  @eisa-rezaei here's the fix -> https://github.com/pmndrs/zustand/pull/2170 you can show your love there hehe. BTW, if you can help us test out the changes using the local instructions would be good so @dai-shi  can merge it as soon as possible

**Reply by alexanderhorner** [reply]
Already tested it. Fixed the issue for me!

**Reply by eisa-rezaei** [reply]
Issue fixed. 👀

**Reply by jwindhager** [reply]
Fixed, thanks so much!

**Comment by j3sch** [comment]
I have the same issue on zustand 4.4.7

**Reply by aaronfulkerson** [reply]
I'm also having this issue now with the following code on 4.5.5:

```
import { create } from "zustand";
import { devtools, persist } from "zustand/middleware";

type State = {
  sourceRole: { id: number; sourceName: string; sourceRole: string } | null;
};

type Actions = {
  setSourceRole: (sourceRole: State["sourceRole"]) => void;
};

export const useSidebarStore = create<State & Actions>(
  devtools(
    persist(
      (set) => ({
        sourceRole: null,
        setSourceRole: (sourceRole) => set(() => ({ sourceRole })),
      }),
      { name: "sidebar-store" }
    )
  )
);
```

Was there a regression? @dbritto-dev

**Reply by dbritto-dev** [reply]
@aaronfulkerson on ts should be `create<T>()(...)` instead of `create<T>(...)`

**Comment by issam-seghir** [comment]
i have similar error 

```ts
Argument of type '{ (nextStateOrUpdater: Store | Partial<Store> | ((state: WritableDraft<Store>) => void), shouldReplace?: false | undefined, action?: Action | undefined): void; (nextStateOrUpdater: Store | ((state: WritableDraft<...>) => void), shouldReplace: true, action?: Action | undefined): void; }' is not assignable to parameter of type '{ (nextStateOrUpdater: CartSlice | Partial<CartSlice> | ((state: WritableDraft<CartSlice>) => void), shouldReplace?: false | undefined, action?: Action | undefined): void; (nextStateOrUpdater: CartSlice | ((state: WritableDraft<...>) => void), shouldReplace: true, action?: Action | undefined): void; }'.
  Types of parameters 'shouldReplace' and 'shouldReplace' are incompatible.
    Type 'true' is not assignable to type 'false'.ts(2345)
```

my store : 
```ts
const useStoreBase = create<Store>()(
	devtools(
		persist(
			subscribeWithSelector(
				immer((...a) => ({
					...createAuthSlice(...a),
					...createCartSlice(...a),
					...createModalSlice(...a),
					...createSidebarSlice(...a),
				}))
			),
			{
				name: "store",
				storage: createJSONStorage(() => AsyncStorage),
			}
		)
	)
);



//? Auto generated selectors for the store using 'auto-zustand-selectors-hook'
//* Usage : useStore.use.selectorName()

export const useStore = createSelectorFunctions(useStoreBase) as typeof useStoreBase & ZustandFuncSelectors<Store>;

//? Setup the devtools using 'simple-zustand-devtools'
//* You can use Redux Devtools Extension as well , with the built in  'zustand/middleware'

if (process.env.NODE_ENV === "development" || __DEV__) {
    mountStoreDevtool("compos-resto-store", useStore);
}

```

**Reply by alexanderhorner** [reply]
Doesn't seem like the problems are related. Did you install the newest version?

**Reply by issam-seghir** [reply]
> Doesn't seem like the problems are related. Did you install the newest version?

i have "zustand": "^5.0.2"

**Reply by mattbieber** [reply]
Getting the same error as @issam-seghir with `immer`.  On each line that spreads a slice inside the immer invocation

**Reply by dbritto-dev** [reply]
@mattbieber that could be expected due to you need to explicitly type your slices but this is not the case

**Comment by daudln** [comment]
Same issue with zustand 5.0.3

**Reply by dbritto-dev** [reply]
ts playground?

**Reply by daudln** [reply]
[TS Playground](https://www.typescriptlang.org/play/?#code/FASwtgDg9gTgLgAgN4IMpwIZwKYGEbZawIC+CAZjFGAgEQBeArgM6YB2AJrQNyiSyIUAYwJZspClRoMW7Lr3DR4yBB2wA3OFCgAbZgBoEEbDGYhWh5owBGzESGvYA6iDgALVNh3YhWmBMpqOiZWDE4AejAQDg5vAHcMAh4+JUEEcDATAKlg2TCOSOjY7ASCcIyTZOBsAA9UhDgAT2MEAEFfEHVXRoQAXmRgBARGRmiALgRWGBA2AHNeIfGENkYwRxgFhAwOrqaASQ4JqZn5wa2d7oA5DEyjuGm5ze24Tu6AEWw7aYgXqDY7h6nIZaCAgIQHAEnTYQHQYRrYQ4IazabxhTZWawAFSgoKEkMeZ3UJhA5BACPxQIaOLBAFVRojjgShgBHRifX5sZgUhAAH2WjB0OnRqzAiUa3L5K0FwrAopgjQAClg3BL+dLgCReNU6gJ0mwcDByNtxDTmFkkGclozKWwbtgKZsWCZbbdJvcoWcMOosIkaTAdKqpUKNcBgE0Wq1GO50GI+gMhk6YBNTVlJQLg0NtkJPsxsQBrbD-N2A3lqjPpZiR9yFl5CMSI5G6QhsXia0Ph8RVtztDnMOMWoZmuBdiYACg4PomA6GwzNSa2bEam0zQmzzFzUALRety4QBEonzc+cLDrOJAAlH0AHwIdRQaKbISomAjhCjy+9G93h8arW1eodm0UYeDoYLiP0XYxjgCAAGRAe4PYgH8zBakIyGIDMrggBgOhQfa8EeJg0H9NOiYTEG+iequObHkWFGWpWwE1mC9YTEaejYJRbbavUaGcogIiEDgkGgdmEx4fgQmwAAPCJYGGAA2gAuopKkEagonYDe-SjmcQ7AB+N6jtOAB0ZmYS8OF4ZRg7YMOwFjhOmCGWctlwMZrkzmRqg+iZiY2TOK5rhuW4TE5GAmVmNGboWAWBeYXbMXWOCIvcbJxaQ54BU+hAvg5b6GZMdmjhZ2G4UR2BZRq55ajxuozAaRrZmguVCG4ABiIA6AafbTrMMAYGop6DjYABWPhwMNVK4qeJChv+9X6iYTXiJ4iRtXh8ZFetKrFh6Qykt1Jhci1O2dUdphPBcLyfNiqBuFAcQTIhuyNMppaMJw2CkmwCKtvNOrKA1y3GqdMBtYhyFbUOa3g7to5mjtFKFd+HDonZ509WOh1YwgSrwGV0mw21mPHVeKP3mjel2S9WG3VA92PWOzyvDdJ2000718p9ag-QiFM-gmEBOdgpPzqOBbiggktQOQYMk11PWGN6Ohssj163pT-11cogHEyBYFxvrm1wfrkOcqh6EIIJYjG5p4kVZJRAwNJZx2-JZwKQpMihBEFQwLQhi-USMBKSpnvh0M7vZsA2lvkOhUebZSN0IHZw48dU6ef1g34bQaeBRi42+BM+cZSCYKlwXpABSzuxkhujNPcMX18xwAUw618OI3DhVDsZ21w5lHcY4rmdvhnpguYXxUIxVicoJPJ0oGZJmhDgJlL4Yq9LxIF5VW5HMN3dD3N6Odd043p998VKAX2zJ+PcPZyMMLYhi2OkvKzhbLTzO-dzzEAvTyB0x6mCnAgVe69sCbzAQYBAClJZKQmCrNkNdPL7y4jVUMfFWBulgBgWY2AADS2Aej9FoKgAA8u1TEAB9TEABRVouAAASjCABKdCAAyVDcCtG4XQ1AmIqEcNaAAcUYckPWfhwLywNs1OCclsx-kBgJK2Tp0CwDkTbHARNZFXnfLpIYahNAomYMYmcxhTDmHciAyYNgvgOGcK4DwXgJqwEsYFRYsoTCjlHKvDAwDvHeNXro7A0dsABLMkEjKISwmiGEsBDSYFomRQPiEzK557EZMCtOEJLp8KsAIUQ0hjQ4lWMSJZUC9B8Lvg1sZC8FS5qBWyUMbJ2CdaIF5hgAUiBNGyN4EAA)

**Reply by dbritto-dev** [reply]
@daudln your types are wrong but here you go -> https://tsplay.dev/w1vPlN

**Reply by daudln** [reply]
> @daudln your types are wrong but here you go -> https://tsplay.dev/w1vPlN

I am good tp goooo!✌️
